cmd/snapd-generator/main.c

```diff
@@ -97,3 +90,80 @@ int main(int argc, char **argv)
 	fprintf(f, "WantedBy=local-fs.target\n");
 	return 0;
 }
+
+int ensure_fusesquashfs_inside_container(const char *normal_dir)
+{
+	// check if we are running inside a container
+	if (system("systemd-detect-virt --container --quiet") != 0) {
```

Comment:
I have a deep seated distrust for system() - there is no user input here and systemd itself calls it with a clean/scrubbed environment so it should actually be fine. But I still feel a bit uneasy :/

Reply:
Mhm.. Can we detect running in a container some other way?

Reply:
you can link with libsystemd0 and call the C api to do the same.

Reply:
ah, but it is not public API... diddums.

Yeah, i don't think there is anything else you can do here.

Reply:
It's fine, if we could use something like glib here we could make this nicer but the alternative of pid=fork(); if (pid==0) execve(...)" and doing the wait etc is also not appealing :/ Because a) no user input b) very controlled environment this should be fine.

Reply:
There is a public D-Bus API: 
```
busctl get-property org.freedesktop.systemd1 /org/freedesktop/systemd1 org.freedesktop.systemd1.Manager Virtualization 
```
Since this is a systemd generator linking with sd-dbus helpers is not out of the question.

Reply:
> There is a public D-Bus API:
> 
> ```
> busctl get-property org.freedesktop.systemd1 /org/freedesktop/systemd1 org.freedesktop.systemd1.Manager Virtualization 
> ```
> 
> Since this is a systemd generator linking with sd-dbus helpers is not out of the question.

no you must not use dbus, as during systemd generator time dbus is not started yet. systemd generators run before any services are started on first boot.

Reply:
The output of container detection is cached in `/run/systemd/container` and that is available before system-generators are executed. So you can check for presence of `/run/systemd/container`.

Reply:
Thanks for pointing this out. I didn't realize this.

Reply:
Sorry for making this such a bike-sheed. Maybe:
```
diff --git a/cmd/snapd-generator/main.c b/cmd/snapd-generator/main.c
index 9e9e7a3714..2e2c948115 100644
--- a/cmd/snapd-generator/main.c
+++ b/cmd/snapd-generator/main.c
@@ -91,10 +91,17 @@ int ensure_root_fs_shared(const char *normal_dir)
        return 0;
 }
 
+static bool file_exists(const char *path) {
+        struct stat buf;
+        return stat(path, &buf) == 0;
+}
+
 int ensure_fusesquashfs_inside_container(const char *normal_dir)
 {
-       // check if we are running inside a container
-       if (system("systemd-detect-virt --container --quiet") != 0) {
+       // check if we are running inside a container, systemd
+        // provides this file all the way back to trusty if run in a
+        // container
+        if (!file_exists("/run/systemd/container")) {
                return 0;
        }
```
to get rid of the first "system()" call ?

Reply:
One typo: return(path, &buf) != 0 <- in file_exists() I think - sorry, wrote this in a hurry.

Reply:
No, I think you original suggestion is fine, 0 is success, thus file exists.

Reply:
Yeah, silly me